File: artifacts/api-server/src/routes/content.ts
import { Router, type IRouter } from "express";
import { db, siteContentTable } from "@workspace/db";
import { eq } from "drizzle-orm";
import { requireAdmin } from "./admin";

const router: IRouter = Router();

const KEY_PATTERN = /^[a-zA-Z0-9._-]+$/;
const MAX_KEY_LEN = 256;
const MAX_VALUE_LEN = 20000;

router.get("/content", async (_req, res) => {
  const rows = await db
    .select({ key: siteContentTable.key, value: siteContentTable.value })
    .from(siteContentTable);
  const map: Record<string, string> = {};
  for (const row of rows) {
    map[row.key] = row.value;
  }
  res.json(map);
});

router.put("/content/:key", requireAdmin, async (req, res) => {
  const key = req.params.key;
  if (!key || key.length > MAX_KEY_LEN || !KEY_PATTERN.test(key)) {
    res.status(400).json({ message: "Invalid key" });
    return;
  }
  const value = typeof req.body?.value === "string" ? req.body.value : null;
  if (value === null) {
    res.status(400).json({ message: "Body must contain a 'value' string" });
    return;
  }
  if (value.length > MAX_VALUE_LEN) {
    res.status(400).json({ message: "Value too long" });
    return;
  }

  const now = new Date();
  await db
    .insert(siteContentTable)
    .values({ key, value, updatedAt: now })
    .onConflictDoUpdate({
      target: siteContentTable.key,
      set: { value, updatedAt: now },
    });

  res.json({ ok: true, key, value });
});

router.delete("/content/:key", requireAdmin, async (req, res) => {
  const key = req.params.key;
  if (!key || key.length > MAX_KEY_LEN || !KEY_PATTERN.test(key)) {
    res.status(400).json({ message: "Invalid key" });
    return;
  }
  await db.delete(siteContentTable).where(eq(siteContentTable.key, key));
  res.json({ ok: true });
});

router.delete("/content", requireAdmin, async (_req, res) => {
  await db.delete(siteContentTable);
  res.json({ ok: true });
});

export default router;

File: artifacts/api-server/app/routers/content.py
import re
from datetime import datetime
from typing import Any, Dict

from fastapi import APIRouter, Depends, Request, status
from fastapi.responses import JSONResponse
from sqlalchemy import delete as sa_delete
from sqlalchemy.dialects.postgresql import insert as pg_insert
from sqlalchemy.orm import Session

from ..auth import require_admin
from ..db import get_db
from ..models import SiteContent

router = APIRouter()

KEY_PATTERN = re.compile(r"^[a-zA-Z0-9._-]+$")
MAX_KEY_LEN = 256
MAX_VALUE_LEN = 20000


def _is_valid_key(key: str) -> bool:
    return bool(key) and len(key) <= MAX_KEY_LEN and bool(KEY_PATTERN.match(key))


@router.get("/content")
def get_content(db: Session = Depends(get_db)) -> Dict[str, str]:
    rows = db.query(SiteContent.key, SiteContent.value).all()
    return {row.key: row.value for row in rows}


@router.put("/content/{key}", dependencies=[Depends(require_admin)])
async def put_content(key: str, request: Request, db: Session = Depends(get_db)):
    if not _is_valid_key(key):
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content={"message": "Invalid key"},
        )

    body: Any
    try:
        body = await request.json()
    except Exception:
        body = None

    value = body.get("value") if isinstance(body, dict) else None
    if not isinstance(value, str):
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content={"message": "Body must contain a 'value' string"},
        )
    if len(value) > MAX_VALUE_LEN:
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content={"message": "Value too long"},
        )

    now = datetime.utcnow()
    stmt = pg_insert(SiteContent).values(key=key, value=value, updated_at=now)
    stmt = stmt.on_conflict_do_update(
        index_elements=[SiteContent.key],
        set_={"value": value, "updated_at": now},
    )
    db.execute(stmt)
    db.commit()
    return {"ok": True, "key": key, "value": value}


@router.delete("/content/{key}", dependencies=[Depends(require_admin)])
def delete_content_key(key: str, db: Session = Depends(get_db)):
    if not _is_valid_key(key):
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content={"message": "Invalid key"},
        )
    db.execute(sa_delete(SiteContent).where(SiteContent.key == key))
    db.commit()
    return {"ok": True}


@router.delete("/content", dependencies=[Depends(require_admin)])
def delete_all_content(db: Session = Depends(get_db)):
    db.execute(sa_delete(SiteContent))
    db.commit()
    return {"ok": True}
